schema.sql

```diff
@@ -0,0 +1,56 @@
+CREATE TABLE users (
+	-- users contains all user-data. An entry in this table corresponds to one
+	-- specific person. The account balance is reconstructed completely out of the
+	-- transactions table, to reduce duplication of information.
+
+
+	-- user_id is a sequential identifier.
+	user_id INTEGER NOT NULL,
+	-- name is the username used for display and login.
+	name TEXT,
+	-- password is a bcrypt-hashed password.
+	password BINARY,
+
+	-- constraints
+	PRIMARY KEY (user_id)
+);
+
+CREATE TABLE cards (
+	-- cards contains all card-data. An entry in this table corresponds to one
+	-- physical card. Every user can have an arbitrary number of cards.
+
+
+	-- card_id is a sequential identifier.
+	card_id BINARY NOT NULL,
```

Comment:
Technisch ist es überlegen, die UID der Karte zu nehmen. Grund: Wir wollen von "An den Leser gehaltene Karte" auf "Owner" matchen. D.h. der lookup muss am Ende eh diese Form haben, wir brauchen eh den index nach UID und wir brauchen auch Eindeutigkeit und NOT NULL und alles, was ein PRIMARY KEY so hat. Eine zusätzliche integer-column einzuführen, ist nur overhead, der dann am Ende nicht benutzt wird (der Grund, warum der primary key so heißt ist, dass es eben der schlüssel ist, der als Hauptidentifikationsmerkmal genutzt und indiziert und entsprechend optimiert wird). Typesafety finde ich kein überzeugendes Argument, denn a) möchte ich mich derzeit noch nicht auf eine SQL-Implementation festlegen und b) ist es egal, *in welcher Spalte* man Typen verhaut, es wird immer ein Problem sein. Die mangelnde typesafety von sqlite spricht eher für BINARY, weil es der native Datentyp ist.

Ein ganz anderes Problem ist, dass die UID tatsächlich eigentlich *kein* sehr gutes eindeutiges Merkmal ist (weshalb der Code bisher sich ja auch auf ISO 14443-A festlegt), nicht jeder NFC-Standard unterstützt so etwas wie eine UID und selbst die, die es tun, tun das nicht unbedingt eindeutig (z.B. die Mifare Karten haben 4, 7 und 10 byte UIDs und insbesondere die 4-byte UIDs (die z.B. von den Wohnheimschlüsseln benutzt werden) garantieren keine uniqueness (weil das gar nicht geht). Aber das ist ein Problem, was ich gerne angehen würde, wenn es soweit ist, weil das echt nicht so leicht zu klären ist (und man kann dann einfach die Karten-Tabelle leeren, ohne kritische Daten zu verlieren). Die wahrscheinlichste Lösung ist, weitere Identifikationsmerkmale auf []byte abzubilden.